File: app/src/main/java/edu/uofk/eeese/eeese/projects/ProjectsFragment.java
/*
 * Copyright 2017 Ali Salah Alddin
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

package edu.uofk.eeese.eeese.projects;

import android.content.Context;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.Snackbar;
import android.support.v4.app.Fragment;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.transitionseverywhere.TransitionManager;

import java.util.List;

import javax.inject.Inject;

import butterknife.BindInt;
import butterknife.BindView;
import butterknife.ButterKnife;
import edu.uofk.eeese.eeese.EEESEapp;
import edu.uofk.eeese.eeese.R;
import edu.uofk.eeese.eeese.data.Project;

/**
 * A fragment that displays a list of projects with a specific ProjectCategory
 * {@link OnProjectSelectedListener} interface to handle project clicks
 */
public class ProjectsFragment extends Fragment implements SwipeRefreshLayout.OnRefreshListener, ProjectsContract.View {

    // Content
    @BindView(R.id.swipe_refresh)
    public SwipeRefreshLayout mSwipeRefresh;
    @BindView(R.id.projects_list)
    public RecyclerView mProjectsList;
    @BindInt(R.integer.number_of_columns)
    public int numOfColumns;
    @BindView(R.id.error_view)
    public View mErrorView;

    //  Control
    private ProjectsAdapter mAdapter;
    private View mSelectedProjectView;
    @Project.ProjectCategory
    private int mCategory;
    private ProjectsAdapter.OnProjectSelectedListener projectSelectedListener =
            new ProjectsAdapter.OnProjectSelectedListener() {
                @Override
                public void showProjectDetails(Project project, View projectCard) {
                    mSelectedProjectView = projectCard;
                    mPresenter.openProjectDetails(project);
                }
            };
    private static final String CATEGORY_KEY = "edu.uofk.eeese.eeese.ProjectsFragment.CATEGORY";

    private OnProjectSelectedListener mListener;

    @Inject
    public ProjectsContract.Presenter mPresenter;

    public ProjectsFragment() {
    }

    public static ProjectsFragment getInstance(@Project.ProjectCategory int category) {
        ProjectsFragment fragment = new ProjectsFragment();
        Bundle arguments = new Bundle();
        arguments.putInt(CATEGORY_KEY, category);
        fragment.setArguments(arguments);
        return fragment;
    }

    /*
     * As always, the int read from the arguments bundle will always be
     * a legal value, because it is always set to a legal value in the first place
     */
    @SuppressWarnings("WrongConstant")
    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Bundle arguments = getArguments();
        if (arguments != null) {
            mCategory = arguments.getInt(CATEGORY_KEY);
        }
        ((EEESEapp) getActivity().getApplicationContext()).getAppComponent()
                .projectsComponent(new ProjectsModule(this, mCategory))
                .inject(this);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        View rootView = inflater.inflate(R.layout.fragment_projects, container, false);

        ButterKnife.bind(this, rootView);

        mSwipeRefresh.setOnRefreshListener(this);

        mAdapter = new ProjectsAdapter(getContext(), projectSelectedListener);
        mProjectsList.setLayoutManager(new LinearLayoutManager(getContext(), LinearLayoutManager.VERTICAL, false));
        mProjectsList.setAdapter(mAdapter);
        mPresenter.loadProjects(false);

        return rootView;
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        if (context instanceof OnProjectSelectedListener) {
            mListener = (OnProjectSelectedListener) context;
        } else {
            throw new RuntimeException(context.toString()
                    + " must implement OnProjectSelectedListener");
        }
    }

    @Override
    public void onResume() {
        super.onResume();
        mPresenter.subscribe();
    }

    @Override
    public void onPause() {
        super.onPause();
        mPresenter.unsubscribe();
    }

    @Override
    public void onDetach() {
        super.onDetach();
        mListener = null;
    }

    @Override
    public void showProjects(@NonNull List<Project> projects) {
        mAdapter = new ProjectsAdapter(getContext(), projects, projectSelectedListener);
        mProjectsList.swapAdapter(mAdapter, true);
        if (mProjectsList.getVisibility() != View.VISIBLE) {
            TransitionManager.beginDelayedTransition(mSwipeRefresh);
        }
        mProjectsList.setVisibility(View.VISIBLE);
        mErrorView.setVisibility(View.GONE);
    }

    @Override
    public void showProjectDetails(@NonNull String projectId) {
        mListener.onProjectSelected(projectId, mSelectedProjectView);
    }

    @Override
    public void showNoProjects() {
        if (mErrorView.getVisibility() != View.VISIBLE) {
            TransitionManager.beginDelayedTransition(mSwipeRefresh);
        }
        mErrorView.setVisibility(View.VISIBLE);
        mProjectsList.setVisibility(View.GONE);

    }

    @Override
    public void setLoadingIndicator(boolean visibility) {
        mSwipeRefresh.setRefreshing(visibility);
    }

    @Override
    public void showNoConnectionError() {
        Snackbar.make(mSwipeRefresh, R.string.connection_error, Snackbar.LENGTH_SHORT).show();
    }

    @Override
    public void setPresenter(@NonNull ProjectsContract.Presenter presenter) {
        mPresenter = presenter;
    }

    @Override
    public void onRefresh() {
        mPresenter.loadProjects(true);
    }

    /**
     * This interface must be implemented by activities that contain this
     * fragment to allow an interaction in this fragment to be communicated
     * to the activity and potentially other fragments contained in that
     * activity.
     * <p>
     * See the Android Training lesson <a href=
     * "http://developer.android.com/training/basics/fragments/communicating.html"
     * >Communicating with Other Fragments</a> for more information.
     */
    public interface OnProjectSelectedListener {
        // TODO: Update argument type and name
        void onProjectSelected(String projectId, View projectView);
    }
}

File: app/src/main/java/edu/uofk/eeese/eeese/projects/ProjectsFragment.kt
/*
 * Copyright 2017 Ali Salah Alddin
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

package edu.uofk.eeese.eeese.projects

import android.content.Context
import android.os.Bundle
import android.support.design.widget.Snackbar
import android.support.v4.app.Fragment
import android.support.v7.widget.GridLayoutManager
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.transitionseverywhere.TransitionManager
import edu.uofk.eeese.eeese.EEESEapp
import edu.uofk.eeese.eeese.R
import edu.uofk.eeese.eeese.data.Project
import edu.uofk.eeese.eeese.data.ProjectCategory
import edu.uofk.eeese.eeese.projects.ProjectsFragment.OnProjectSelectedListener
import edu.uofk.eeese.eeese.util.OffsetItemDecorator
import io.reactivex.disposables.Disposable
import io.reactivex.disposables.Disposables
import kotlinx.android.synthetic.main.fragment_projects.*
import javax.inject.Inject


/**
 * A fragment that displays a list of projects with a specific ProjectCategory
 * [OnProjectSelectedListener] interface to handle project clicks
 */
class ProjectsFragment : Fragment(), ProjectsContract.View {
    companion object {
        private val CATEGORY_KEY = "edu.uofk.eeese.eeese.ProjectsFragment.CATEGORY"

        private const val CATEGORY_SOFTWARE = 0
        private const val CATEGORY_POWER = 1
        private const val CATEGORY_TELECOM = 2
        private const val CATEGORY_ELECTRONICS_CONTROL = 3

        fun getInstance(category: ProjectCategory): ProjectsFragment {
            val fragment = ProjectsFragment()
            val arguments = Bundle()
            arguments.putInt(CATEGORY_KEY,
                    when (category) {
                        ProjectCategory.SOFTWARE -> CATEGORY_SOFTWARE
                        ProjectCategory.POWER -> CATEGORY_POWER
                        ProjectCategory.TELECOM -> CATEGORY_TELECOM
                        ProjectCategory.ELECTRONICS_CONTROL -> CATEGORY_ELECTRONICS_CONTROL
                    })
            fragment.arguments = arguments
            return fragment
        }

        private fun category(category: Int) = when (category) {
            CATEGORY_SOFTWARE -> ProjectCategory.SOFTWARE
            CATEGORY_POWER -> ProjectCategory.POWER
            CATEGORY_TELECOM -> ProjectCategory.TELECOM
            CATEGORY_ELECTRONICS_CONTROL -> ProjectCategory.ELECTRONICS_CONTROL
            else -> throw IllegalArgumentException("unknown category code '$category'")
        }
    }

    private lateinit var category: ProjectCategory

    private var projectClicks: Disposable = Disposables.disposed()
    private var selectedProject: View? = null

    lateinit private var projectsAdapter: ProjectsAdapter

    private var mListener: OnProjectSelectedListener? = null

    @Inject lateinit var projectsPresenter: ProjectsContract.Presenter

    /*
     * As always, the int read from the arguments bundle will always be
     * a legal value, because it is always set to a legal value in the first place
     */
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if (arguments != null) {
            category = category(arguments.getInt(CATEGORY_KEY))
        }
        (activity.applicationContext as EEESEapp).appComponent
                .projectsComponent(ProjectsModule(this, category))
                .inject(this)
    }

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        // Inflate the layout for this fragment
        val rootView = inflater!!.inflate(R.layout.fragment_projects, container, false)

        // to identify the fragment with its category
        rootView.setTag(R.string.tag_projectslist, projectsPresenter.getCategory())
        return rootView
    }

    override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        swipe_refresh.setOnRefreshListener { reloadProjects() }
        swipe_refresh.setColorSchemeResources(R.color.colorAccent)

        projectsAdapter = ProjectsAdapter()

        val numberOfColumns = resources.getInteger(R.integer.number_of_columns)
        projects_list.layoutManager = GridLayoutManager(context, numberOfColumns)
        projects_list.adapter = projectsAdapter
        projects_list.addItemDecoration(OffsetItemDecorator(
                resources.getDimensionPixelSize(R.dimen.project_margin_vertical),
                resources.getDimensionPixelSize(R.dimen.project_margin_horizontal),
                numberOfColumns))


        reload_button.setOnClickListener { reloadProjects() }
        projectsPresenter.loadProjects(false)

    }

    override fun onAttach(context: Context?) {
        super.onAttach(context)
        if (context is OnProjectSelectedListener?) {
            mListener = context
        } else {
            throw RuntimeException(context!!.toString() + " must implement OnProjectSelectedListener")
        }
    }

    override fun onResume() {
        super.onResume()
        projectsPresenter.subscribe()
        registerProjectClicks()
    }

    override fun onPause() {
        super.onPause()
        projectsPresenter.unsubscribe()
        projectClicks.dispose()
    }

    override fun onDetach() {
        super.onDetach()
        mListener = null
    }

    override fun showProjects(projects: List<Project>) {
        projectsAdapter = ProjectsAdapter(projects)
        projectClicks.dispose()
        registerProjectClicks()

        projects_list.swapAdapter(projectsAdapter, false)
        if (projects_list.visibility != View.VISIBLE) {
            TransitionManager.beginDelayedTransition(swipe_refresh)
        }
        projects_list.visibility = View.VISIBLE
        error_view.visibility = View.GONE
    }

    override fun showProjectDetails(projectId: String) {
        mListener!!.onProjectSelected(projectId, selectedProject!!)
    }

    override fun showNoProjects() {
        if (error_view.visibility != View.VISIBLE) {
            TransitionManager.beginDelayedTransition(swipe_refresh)
        }
        error_view.visibility = View.VISIBLE
        projects_list.visibility = View.GONE

    }

    override fun setLoadingIndicator(visibility: Boolean) {
        swipe_refresh.isRefreshing = visibility
    }

    override fun showConnectionError() {
        Snackbar.make(swipe_refresh, R.string.connection_error, Snackbar.LENGTH_SHORT).show()
    }

    override fun setPresenter(presenter: ProjectsContract.Presenter) {
        projectsPresenter = presenter
    }

    fun reloadProjects() {
        projectsPresenter.loadProjects(true)
    }

    private fun registerProjectClicks() {
        projectClicks = projectsAdapter
                .projectClicks()
                .subscribe {
                    selectedProject = it.projectCard
                    projectsPresenter.openProjectDetails(it.project)
                }
    }

    /**
     * This interface must be implemented by activities that contain this
     * fragment to allow an interaction in this fragment to be communicated
     * to the activity and potentially other fragments contained in that
     * activity.
     *
     */
    interface OnProjectSelectedListener {
        fun onProjectSelected(projectId: String, projectView: View)
    }
}
